name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/QuirionPDFExtractor.java

```diff
@@ -0,0 +1,377 @@
+        type.addBlock(dividendBlock);
+        dividendBlock.set(new Transaction<AccountTransaction>().subject(() -> {
+            AccountTransaction t = new AccountTransaction();
+            t.setType(AccountTransaction.Type.DIVIDENDS);
+            return t;
+        })
+
+                        .section("isin", "name").match("^Erträgnisabrechnung, Ref.: \\d+ (\\d+.\\d+.\\d{4}+) (\\d+.\\d+.\\d{4}+) ([\\d.]+,\\d{2})(.*)$")
+                        .match("^(?<name>.*)$").match("((^Dis.oN\\n)|(^))(?<isin>.{12})(.*)").assign((t, v) -> {
+                            Map<String, String> context = type.getCurrentContext();
+                            v.put("currency", context.get("currency"));
+                            t.setSecurity(getOrCreateSecurity(v));
+                        })
+
+                        .section("amount", "date", "shares")
+                        .match("^Erträgnisabrechnung, Ref.: \\d+ (\\d+.\\d+.\\d{4}+) (?<date>\\d+.\\d+.\\d{4}+) (?<amount>[\\d.]+,\\d{2})(.*)$")
+                        .match("^(.*)$")
+                        .match("((^Dis.oN\\n)|(^))(.{12}, ST) (?<shares>[\\d.,]+)(.*)")
+                        .assign((t, v) -> {
+                            Map<String, String> context = type.getCurrentContext();
+                            t.setDateTime(asDate(v.get("date")));
+                            t.setShares(asShares(v.get("shares")));
+                            t.setCurrencyCode(asCurrencyCode(context.get("currency")));
+                            t.setAmount(asAmount(v.get("amount")));
+                        })
+                        .section("tax", "currency").optional()
+                        .match("^Erträgnisabrechnung, Ref.: \\d+ (\\d+.\\d+.\\d{4}+) (?<date>\\d+.\\d+.\\d{4}+) (?<amount>[\\d.]+,\\d{2})(.*)$")
+                        .match("^(.*)$")
+                        .match("^((Dis.oN\\n)|())(.{12}, ST) ([\\d.,]+)(.*)$")
+                        .match("^(.*)(KEST: )(?<currency>\\w{3}) -(?<tax>[\\d.]+,\\d{2}).*")
+                        .assign((t, v) -> {//t.addUnit(new Unit(Unit.Type.TAX, asAmount(v.get("tax"))));
+                                            Money taxes = Money.of(asCurrencyCode("EUR"),
+  //TODO other currencys not working                         Money.of(asCurrencyCode(v.get("currency")), asAmount(v.get("tax"))))))
+                                                            asAmount(v.get("tax")));
+                                            //.subtract(t.getMonetaryAmount()).toMoney();
+                                          t.addUnit(new Unit(Unit.Type.TAX, taxes));})
+                        //.assign((t, v) -> t.addUnit(new Unit(Unit.Type.TAX, asAmount(v.get("tax")))))
+//                                        Money.of(asCurrencyCode(v.get("currency")), asAmount(v.get("tax"))))))
+
+                        .section("tax", "currency").optional()
```

Comment:
Wouldn't it make sense to read KEST and SOLI at the same time because it is always printed in one line?

Something like:

`.section("kest", "soli", "currency").optional()`
...
`.match("^KEST: (?<currency>\\w{3}) -(?<tax>[\\d\\.]+,\\d{2}), SOLI: (?<currency>\\w{3}) -(?<tax>[\\d\\.]+,\\d{2})(.*)$")`

Then reading both values and adding them up.

Reply:
please see example in Kontoauszug01.txt line 80. If KSST is very low SOLI is not in the text.
`KEST: EUR -0,01`

Reply:
Is this regex better? I don't know, if it passes back an empty value for soli, in case it is not given.

`^KEST: (?<currency>\w{3}) -(?<kest>[\d\.]+,\d{2})(, SOLI: (\w{3}) -)?((?<soli>[\d\.]+,\d{2})|(.*))?$`
 
Otherwise, a `oneof` statement might make sense?!

Reply:
I believe both approaches are valid - match one line in a block multiple times (once for KEST and once for SOLI) or match it with one block.

the ```oneOf```  would only work if we are guaranteed to always have at least KEST or SOLI.

the problem I have is that there are no exchanges rates in the document. See below